libethereum/BlockChain.cpp

```diff
@@ -612,31 +599,33 @@ void BlockChain::insert(VerifiedBlockRef _block, bytesConstRef _receipts, bool _
 			m_details[_block.info.parentHash()].children.push_back(_block.info.hash());
 	}
 
-	blocksBatch.Put(toSlice(_block.info.hash()), ldb::Slice(_block.block));
+	blocksTransaction->insert(toSlice(_block.info.hash()), db::Slice(_block.block));
 	DEV_READ_GUARDED(x_details)
-		extrasBatch.Put(toSlice(_block.info.parentHash(), ExtraDetails), (ldb::Slice)dev::ref(m_details[_block.info.parentHash()].rlp()));
+		extrasTransaction->insert(toSlice(_block.info.parentHash(), ExtraDetails), (db::Slice)dev::ref(m_details[_block.info.parentHash()].rlp()));
 
 	BlockDetails bd((unsigned)pd.number + 1, pd.totalDifficulty + _block.info.difficulty(), _block.info.parentHash(), {});
-	extrasBatch.Put(toSlice(_block.info.hash(), ExtraDetails), (ldb::Slice)dev::ref(bd.rlp()));
-	extrasBatch.Put(toSlice(_block.info.hash(), ExtraLogBlooms), (ldb::Slice)dev::ref(blb.rlp()));
-	extrasBatch.Put(toSlice(_block.info.hash(), ExtraReceipts), (ldb::Slice)_receipts);
+	extrasTransaction->insert(toSlice(_block.info.hash(), ExtraDetails), (db::Slice)dev::ref(bd.rlp()));
+	extrasTransaction->insert(toSlice(_block.info.hash(), ExtraLogBlooms), (db::Slice)dev::ref(blb.rlp()));
+	extrasTransaction->insert(toSlice(_block.info.hash(), ExtraReceipts), (db::Slice)_receipts);
 
-	ldb::Status o = m_blocksDB->Write(m_writeOptions, &blocksBatch);
-	if (!o.ok())
+	try
+	{
+		blocksTransaction->commit();
```

Comment:
I don't like this design where you have to wrap every DB call with a try-catch block. If using exceptions in the API does not allow you to have more generic exception handler this does not make much sense.

Reply:
Yes I agree with you to an extent although I followed the perhaps outdated style guide that suggested using exceptions instead of error values where possible.

Reply:
That's a valid suggestion, but don't add try-catch around every call, make a more general try-catch somewhere above for all of the throwing calls (catching `Exception` probably)

Here to distinguish between writing to blocks and extras you could have two different exception classes

Reply:
Honestly, not sure how this is much different than the `if`/`else` that was here originally. The idea of exiting on commit failure still has to be explicit.

Reply:
The whole `BlockChain::insert` method can have probably a single try/catch, a single concise error handler probably. That's very different from if/else for each call.

Reply:
The original implementation was a bunch of if/else statements which are now try/catch statements. I like your suggestion for a large unified try/catch block. Just wondering if the if/else was acceptable, and if so, why multiple try/catch is any worse.

Reply:
The original I guess was the old code not really in compliance with current coding standards